ts options resolving

```
> swr@2.2.4 build-package /Users/huozhi/workspace/swr
> bunchee --cwd "_internal"

error TS5074: Option '--incremental' can only be specified using tsconfig, emitting to single file or when option '--tsBuildInfoFile' is specified.
```

`tsBuildInfoFile` should use default value when incremental is true or disable incremental for build

--- a/src/build-config.ts
+++ b/src/build-config.ts
@@ -67,14 +67,14 @@ function getBuildEnv(envs: string[]) {
   return envVars
 }
 
-function buildInputConfig(
+async function buildInputConfig(
   entry: string,
   pkg: PackageMetadata,
   options: BundleOptions,
   cwd: string,
   { tsConfigPath, tsCompilerOptions }: TypescriptOptions,
   dts: boolean,
-): InputOptions {
+): Promise<InputOptions> {
   const hasNoExternal = options.external === null
   const externals = hasNoExternal
     ? []
@@ -124,9 +124,9 @@ function buildInputConfig(
   // common plugins for both dts and ts assets that need to be processed
   const commonPlugins = [sizePlugin]
 
-  let baseResolvedTsOptions
+  let baseResolvedTsOptions: TypescriptOptions['tsCompilerOptions'] | undefined
   if (dts && useTypescript) {
-    baseResolvedTsOptions = convertCompilerOptions(cwd, {
+    baseResolvedTsOptions = (await convertCompilerOptions(cwd, {
       declaration: true,
       noEmit: false,
       noEmitOnError: true,
@@ -139,7 +139,7 @@ function buildInputConfig(
       module: 'esnext',
       incremental: false,
       jsx: tsCompilerOptions.jsx || 'react',
-    })
+    })).options
   }
 
   const plugins: Plugin[] = (
@@ -363,7 +363,7 @@ export async function buildEntryConfig(
   return (await Promise.all(configs)).filter(nonNullable)
 }
 
-function buildConfig(
+async function buildConfig(
   entry: string,
   pkg: PackageMetadata,
   exportPaths: ExportPaths,
@@ -372,11 +372,11 @@ function buildConfig(
   cwd: string,
   tsOptions: TypescriptOptions,
   dts: boolean,
-): BuncheeRollupConfig {
+): Promise<BuncheeRollupConfig> {
   const { file } = bundleConfig
   const useTypescript = Boolean(tsOptions.tsConfigPath)
   const options = { ...bundleConfig, useTypescript }
-  const inputOptions = buildInputConfig(
+  const inputOptions = await buildInputConfig(
     entry,
     pkg,
     options,

--- a/src/build-config.ts
+++ b/src/build-config.ts
@@ -8,9 +8,8 @@ import type {
   FullExportCondition,
 } from './types'
 import type { InputOptions, OutputOptions, Plugin } from 'rollup'
-import type { TypescriptOptions } from './typescript'
-
-import { convertCompilerOptions } from './typescript'
+import { convertCompilerOptions, type TypescriptOptions } from './typescript'
+import type { CompilerOptions } from 'typescript'
 import { resolve, dirname } from 'path'
 import { wasm } from '@rollup/plugin-wasm'
 import { swc } from 'rollup-plugin-swc3'
@@ -91,7 +90,7 @@ async function buildInputConfig(
     minify: shouldMinify,
   } = options
   const hasSpecifiedTsTarget = Boolean(
-    tsCompilerOptions?.target && tsConfigPath,
+    tsCompilerOptions.target && tsConfigPath,
   )
 
   const swcParserConfig = {
@@ -124,38 +123,44 @@ async function buildInputConfig(
   // common plugins for both dts and ts assets that need to be processed
   const commonPlugins = [sizePlugin]
 
-  let baseResolvedTsOptions: TypescriptOptions['tsCompilerOptions'] | undefined
-  if (dts && useTypescript) {
-    baseResolvedTsOptions = (await convertCompilerOptions(cwd, {
-      declaration: true,
-      noEmit: false,
-      noEmitOnError: true,
-      emitDeclarationOnly: true,
-      checkJs: false,
-      declarationMap: false,
-      skipLibCheck: true,
-      preserveSymlinks: false,
+  const baseResolvedTsOptions: CompilerOptions = {
+    declaration: true,
+    noEmit: false,
+    noEmitOnError: true,
+    emitDeclarationOnly: true,
+    checkJs: false,
+    declarationMap: false,
+    skipLibCheck: true,
+    preserveSymlinks: false,
+    incremental: false,
+  }
+
+  const typesPlugins = [
+    ...commonPlugins,
+    inlineCss({ skip: true }),
+  ]
+
+  if (useTypescript) {
+    const { options: resolvedTsOptions } = await convertCompilerOptions(cwd, {
+      ...baseResolvedTsOptions,
       target: 'esnext',
       module: 'esnext',
-      incremental: false,
       jsx: tsCompilerOptions.jsx || 'react',
-    })).options
+    })
+    const dtsPlugin = (require('rollup-plugin-dts') as typeof import('rollup-plugin-dts')).default({
+      tsconfig: tsConfigPath,
+      compilerOptions: {
+        ...tsCompilerOptions,
+        ...resolvedTsOptions,
+      },
+    })
+    typesPlugins.push(dtsPlugin)
   }
+
 
   const plugins: Plugin[] = (
     dts
-      ? [
-          ...commonPlugins,
-          inlineCss({ skip: true }),
-          useTypescript &&
-            require('rollup-plugin-dts').default({
-              tsconfig: tsConfigPath,
-              compilerOptions: {
-                ...tsCompilerOptions,
-                ...baseResolvedTsOptions,
-              },
-            }),
-        ]
+      ? typesPlugins
       : [
           ...commonPlugins,
           inlineCss({ exclude: /node_modules/ }),

--- a/src/build-config.ts
+++ b/src/build-config.ts
@@ -8,7 +8,7 @@ import type {
   FullExportCondition,
 } from './types'
 import type { InputOptions, OutputOptions, Plugin } from 'rollup'
-import { type TypescriptOptions } from './typescript'
+import { convertCompilerOptions, type TypescriptOptions } from './typescript'
 
 import { resolve, dirname } from 'path'
 import { wasm } from '@rollup/plugin-wasm'
@@ -132,7 +132,10 @@ async function buildInputConfig(
     declarationMap: false,
     skipLibCheck: true,
     preserveSymlinks: false,
+    // disable incremental build
     incremental: false,
+    // use default tsBuildInfoFile value
+    tsBuildInfoFile: '.tsbuildinfo',
     target: 'esnext',
     module: 'esnext',
     jsx: tsCompilerOptions.jsx || 'react',
@@ -144,16 +147,21 @@ async function buildInputConfig(
   ]
 
   if (useTypescript) {
+    const mergedOptions = {
+      ...baseResolvedTsOptions,
+      ...tsCompilerOptions,
+    }
+
+    // error TS5074: Option '--incremental' can only be specified using tsconfig, emitting to single
+    // file or when option '--tsBuildInfoFile' is specified.
+    delete mergedOptions.incremental
+
     const dtsPlugin = (require('rollup-plugin-dts') as typeof import('rollup-plugin-dts')).default({
-      tsconfig: tsConfigPath,
-      compilerOptions: {
-        ...baseResolvedTsOptions,
-        ...tsCompilerOptions,
-      },
+      tsconfig: undefined,
+      compilerOptions: mergedOptions,
     })
     typesPlugins.push(dtsPlugin)
   }
-
 
   const plugins: Plugin[] = (
     dts

--- a/src/build-config.ts
+++ b/src/build-config.ts
@@ -8,8 +8,8 @@ import type {
   FullExportCondition,
 } from './types'
 import type { InputOptions, OutputOptions, Plugin } from 'rollup'
-import { convertCompilerOptions, type TypescriptOptions } from './typescript'
-import type { CompilerOptions } from 'typescript'
+import { type TypescriptOptions } from './typescript'
+
 import { resolve, dirname } from 'path'
 import { wasm } from '@rollup/plugin-wasm'
 import { swc } from 'rollup-plugin-swc3'
@@ -123,7 +123,7 @@ async function buildInputConfig(
   // common plugins for both dts and ts assets that need to be processed
   const commonPlugins = [sizePlugin]
 
-  const baseResolvedTsOptions: CompilerOptions = {
+  const baseResolvedTsOptions: any = {
     declaration: true,
     noEmit: false,
     noEmitOnError: true,
@@ -133,6 +133,9 @@ async function buildInputConfig(
     skipLibCheck: true,
     preserveSymlinks: false,
     incremental: false,
+    target: 'esnext',
+    module: 'esnext',
+    jsx: tsCompilerOptions.jsx || 'react',
   }
 
   const typesPlugins = [
@@ -141,17 +144,11 @@ async function buildInputConfig(
   ]
 
   if (useTypescript) {
-    const { options: resolvedTsOptions } = await convertCompilerOptions(cwd, {
-      ...baseResolvedTsOptions,
-      target: 'esnext',
-      module: 'esnext',
-      jsx: tsCompilerOptions.jsx || 'react',
-    })
     const dtsPlugin = (require('rollup-plugin-dts') as typeof import('rollup-plugin-dts')).default({
       tsconfig: tsConfigPath,
       compilerOptions: {
+        ...baseResolvedTsOptions,
         ...tsCompilerOptions,
-        ...resolvedTsOptions,
       },
     })
     typesPlugins.push(dtsPlugin)

--- a/src/build-config.ts
+++ b/src/build-config.ts
@@ -8,7 +8,7 @@ import type {
   FullExportCondition,
 } from './types'
 import type { InputOptions, OutputOptions, Plugin } from 'rollup'
-import { convertCompilerOptions, type TypescriptOptions } from './typescript'
+import { type TypescriptOptions } from './typescript'
 
 import { resolve, dirname } from 'path'
 import { wasm } from '@rollup/plugin-wasm'
@@ -154,7 +154,10 @@ async function buildInputConfig(
 
     // error TS5074: Option '--incremental' can only be specified using tsconfig, emitting to single
     // file or when option '--tsBuildInfoFile' is specified.
-    delete mergedOptions.incremental
+    if (!mergedOptions.incremental) {
+      delete mergedOptions.incremental
+      delete mergedOptions.tsBuildInfoFile
+    }
 
     const dtsPlugin = (require('rollup-plugin-dts') as typeof import('rollup-plugin-dts')).default({
       tsconfig: undefined,

--- a/test/integration.test.ts
+++ b/test/integration.test.ts
@@ -1,4 +1,5 @@
 import fs from 'fs/promises'
+import fsSync from 'fs'
 import { execSync, fork } from 'child_process'
 import { resolve, join } from 'path'
 import { stripANSIColor, existsFile, assertFilesContent } from './testing-utils'
@@ -221,6 +222,24 @@ const testCases: {
       expect(await fs.readFile(distFiles[1], 'utf-8')).toContain(
         'declare function _default(): string;',
       )
+    },
+  },
+  {
+    name: 'ts-incremental',
+    args: [],
+    async expected(dir) {
+      const distFiles = [
+        './dist/index.js',
+        './dist/index.d.ts',
+      ]
+
+      for (const f of distFiles) {
+        expect(await existsFile(join(dir, f))).toBe(true)
+      }
+      expect(await fs.readFile(join(dir, distFiles[1]), 'utf-8')).toContain(
+        'declare const _default: () => string;',
+      )
+      expect(await existsFile(join(dir, './dist/.tsbuildinfo'))).toBe(false)
     },
   },
   {

--- a/test/integration/ts-incremental/package.json
+++ b/test/integration/ts-incremental/package.json
@@ -1,0 +1,5 @@
+{
+  "name": "ts-incremental",
+  "types": "./dist/index.d.ts",
+  "exports": "./dist/index.js"
+}

--- a/test/integration/ts-incremental/src/index.ts
+++ b/test/integration/ts-incremental/src/index.ts
@@ -1,0 +1,1 @@
+export default () => 'index'

--- a/test/integration/ts-incremental/tsconfig.json
+++ b/test/integration/ts-incremental/tsconfig.json
@@ -1,0 +1,5 @@
+{
+  "compilerOptions": {
+    "incremental": true
+  }
+}

